### ISSUE TITLE
Type hinting take too long to be resolved

### ISSUE THREAD

**Issue by Tomperez98** [microsoft/pylance-release#4243]
## Environment data

-   Language Server version: Pylance v2023.4.20
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.9.16 - pip

## Code Snippet

```python
`__main__.py`
inputs = RawChatInput(user_prompt=prompt)
workflow = RawChat(inputs=inputs)
workflow_result = workflow.execute(
    write_ops_registry=UseCaseRegistry[Any](
        max_length=0,
    ),
)

`other_file.py`
class RawChat(BaseCommand):

    def __init__(self, inputs: RawChatAIInput) -> None:
        self.inputs = inputs

    def execute(
        self,
        write_ops_registry: UseCaseRegistry[Any],
    ) -> Result[str, Union[CommandInputValidationError, UseCaseExecutionError]]:
        ...
```

## Expected behavior

Type hint for `workflow_result` should be of type `Result[str, Union[CommandInputValidationError, UseCaseExecutionError]]`

## Actual behavior

Pylance cannot resolve the type annotation and displays `Any`
I'm using this library right here https://pypi.org/project/result/

This issue only started happening after I installed several libraries, and after quite an amount of time it starts showing the type annotation correctly. Maybe I'm not configuring `Pylance` to correctly ignore `.venv/`, idk

**Comment by Tomperez98** [comment]
```json
{
    "window.zoomLevel": 0.3,
    "editor.minimap.enabled": true,
    "editor.renderWhitespace": "all",
    "editor.formatOnSave": true,
    "editor.rulers": [
        88,
        100
    ],
    "python.testing.pytestArgs": [
        "tests",
        // https://github.com/microsoft/vscode-python/issues/693#issuecomment-367856613
        "--no-cov",
    ],
    "[python]": {
        "gitlens.codeLens.symbolScopes": [
            "!Module"
        ],
        "editor.wordBasedSuggestions": true,
    },
    "python.linting.lintOnSave": true,
    "python.linting.enabled": true,
    "python.formatting.provider": "black",
    "python.linting.mypyEnabled": true,
    "python.testing.pytestEnabled": true,
    "python.analysis.typeCheckingMode": "off",
    "python.languageServer": "Pylance",
    "python.analysis.diagnosticMode": "openFilesOnly",
    "python.analysis.autoSearchPaths": false,
    "python.analysis.useLibraryCodeForTypes": true // Mypy is doing this.
}
```

This is my settings

**Comment by erictraut** [comment]
Could you provide a self-contained minimal code sample? The sample you've provided above refers to a bunch of symbols that are not builtins, and you don't show how they are defined or imported.

**Comment by erictraut** [comment]
Also, it's not clear to me from your bug report whether you're seeing unexpected type evaluations or whether the type evaluation is expected but it's taking a long time to report it (i.e. you're seeing a "Loading" message for a while when hovering over the symbol). Could you clarify which of the two you're experiencing?

Or are you saying that the type evaluation seems to change over time — that it is initially shown as `Any` but later changes to `Result[...]`, even though you've made no changes in your code?

**Comment by Tomperez98** [comment]
Ok. You're right, but how does having a library (maybe a heavy one) affects the Pylance load time?

**Comment by Tomperez98** [comment]
<img width="969" alt="image" src="https://user-images.githubusercontent.com/72174660/232140389-3ab1b39c-d257-46a8-b107-eacdc994693f.png">

This is the weird part. The `reveal_type` function from `typing_extensions` resolves the type hint while Pylance still reports Any

**Comment by Tomperez98** [comment]
This might be irrelevant but as soon as I do a save command on the `pyproject.toml` definition everything works as expected. But I need to explicitly open the `pyproject.toml` and do save (no changes at all)

Has there been any similar issues?

**Comment by erictraut** [comment]
Saving the pyproject.toml forces pylance to discard all type information it has previously evaluated and recompute all types from scratch.

If you're able to provide repro steps with a self-contained minimal code sample, we can investigate further.

**Comment by Tomperez98** [comment]
```python
from result import Err, Ok, Result
from typing_extensions import reveal_type


def divide(a: float, b: float) -> Result[float, ZeroDivisionError]:
    try:
        c = a / b
    except ZeroDivisionError as e:
        return Err(e)
    return Ok(c)


if __name__ == "__main__":
    division_result = divide(a=1.7, b=2)
    reveal_type(division_result)

```

**Comment by erictraut** [comment]
Thanks, that is a self-contained minimal code sample. However, I'm not seeing any unexpected behaviors here. What behavior are you seeing with this sample? Are you seeing a performance issue? Unexpected type evaluations?

**Comment by Tomperez98** [comment]
Maybe it's my setup or the size of the libraries I have installed. But, for me, the type annotation for the variable `division_result` is `Any` at first... then, after saving `pyproject.toml` I get `Result[float, ZeroDivisionError]` (which is the expected)

**Comment by Tomperez98** [comment]
So the question is why VSCode is not picking `pyproject.toml` at first

**Comment by erictraut** [comment]
When you say "the type annotation for the variable is `Any` at first", do you mean that the revealed type output by the `reveal_type` call is `Any`?

Here's what I see when I paste your code into VS Code with Pylance installed.

![Screenshot 2023-04-14 at 1 53 02 PM](https://user-images.githubusercontent.com/7040122/232152536-3a2f6299-3510-4c7c-812d-3a889b1a1003.png)

As I mentioned, when Pylance detects a change to `pyproject.toml`, it discards any cached type information it has already evaluated and re-evaluates everything from scratch. You can also force Pylance to restart itself by running the "Python: Restart Language Server" command from the command palette. Does this affect the revealed type?

Can you repro the problem with a newly-created project and/or a newly-created Python virtual environment?

**Comment by Tomperez98** [comment]
That's right:
<img width="639" alt="image" src="https://user-images.githubusercontent.com/72174660/232157496-2b1fe5f3-6912-4ed3-9397-8ceb942a3ef7.png">


I guess is do to some dependencies that make the load process slow at first. Idk.

Either way. `touch pyproject.toml` solved the issue. It's a workaround I guess but it does the job just fine.

**Comment by erictraut** [comment]
> I guess is do to some dependencies that make the load process slow at first

No, this is unlikely to have anything to do with performance or project size. Type evaluations should be stable and consistent regardless of performance. If it were slow, you would simply need to wait a while before the diagnostic appeared for the `reveal_type` statement. Once this diagnostic appears, it should display the correct type information every time. So this definitely appears to be a bug. However, without a way to repro it or some additional clues, we don't have much to go on.

Please add the following lines below the existing `reveal_type` statement. What types do these reveal?
```python
    reveal_type(Err)
    reveal_type(Ok)
    reveal_type(Result)
```

Did the "Python: Restart Language Server" command cause the revealed type to change from `Any`?

**Comment by Tomperez98** [comment]
<img width="688" alt="image" src="https://user-images.githubusercontent.com/72174660/232163871-0f5f0faf-f786-442f-8769-17e6c5f30677.png">

After reloading `pyproject.toml`

<img width="870" alt="image" src="https://user-images.githubusercontent.com/72174660/232163919-71f8582a-2fbe-41cd-877b-821bcc88b939.png">

**Comment by erictraut** [comment]
One suspicious thing I see in the `result.py` module is the following:
```python
if sys.version_info[:2] >= (3, 10):
    from typing import Final, Literal, ParamSpec, TypeAlias
else:
    from typing_extensions import Final, Literal, ParamSpec, TypeAlias
```

This will work at runtime, but it will generally cause problems for static analysis tools because it's using a non-standard way to perform a version check. The standard way (and the pattern that is supported in PEP 484 by static type checkers) is:
```python
if sys.version_info >= (3, 10): ...
```

Because this library isn't using a standard expression pattern here, the type evaluator cannot statically determine which of these two code paths will be taken at runtime, so it needs to evaluate the type of `TypeAlias` for both code paths and return the union of the two types. This should still work fine (and it does on my system) because `TypeAlias` should be the same type in both the `typing` and the `typing_extensions` stubs. However, it's possible that you have some package installed in your Python environment that is overriding the stubs that are bundled with pylance. 

Try this... Within your Python environment, run `pip uninstall typing_extensions-stubs`. There was a version of this package that was errantly published to pypi for a short period of time, and I wonder if you installed it in your environment.

If that doesn't explain the problem, then the next thing I recommend you try is to create a fresh Python environment. You can create a new virtual environment from the terminal by typing `python3.9 -m venv <path of new virtual environment>`. Then activate this virtual environment and run `pip install result` to install the `result` library and its dependencies. Then create a fresh new project in VS Code and select the new venv. Create a new "test.py" source file and paste in your code sample above. Are you able to repro the problem now?

**Comment by Tomperez98** [comment]
Thanks! @erictraut
